### ISSUE TITLE
Add support for CropAndResize

### ISSUE THREAD

**Issue by dsmilkov** [tensorflow/tfjs#624]
We should add mapping for `CropAndResize` in the Converter. See [discussion](https://groups.google.com/a/tensorflow.org/forum/?utm_medium=email&utm_source=footer#!msg/tfjs/lqLhRGpvm0A/2hjYpLRXAQAJ) for context.

**Comment by Yorkchung** [comment]
So , how to add it? I have no sense.

**Comment by AviKndr** [comment]
@dsmilkov @pyu10055 Hi, I'm currently trying to implement `CropAndResize`, but this is my first contribution so I would appreciate any advice you could provide me. I've already implemented the cpu forward and backprop kernel functions in `backend_cpu.ts` and `backend.ts`, and the op in `image_ops.ts`. I had a few questions about correct and efficient implementation. I have based my implementation on [https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/kernels/crop_and_resize_op.cc](url)

You can find mine here: [https://github.com/AviKndr/tfjs-core](url)

1. From my understanding, `get()` calls in for loops are inefficient for tensors stored in gpu's as they make calls to `dataSync()` repetitively for the whole tensor. I manually calculated flat indexes within the nested for loops and indexed a `TypedArray` instead. If this is correct reasoning, I have noticed that resizeBilinear and resizeNearestNeighbor make excessive use of get() calls and I can change them too. 

2. The op only back-propagates box gradients when the resize method is `bilinear` so I created two backwards functions for the op in `image_ops.ts` and picked which one I passed to `runKernel` based on the method. I am not sure if this is the preferred method so that's why I am asking. 

3. I realize I can add tests in `image_ops_test.ts` but the tests in `ops/image_ops_test.cc` in the regular tensorflow repo are mostly for correct input and output shapes. How do I test correctness of the actual output values. Is there a way to test directly against python tensorflow?

3. I am new to webgl programming so any advice on resources to quickly pick up understanding of cuda and webgl so I can translate the gpu backend version would be appreciated. 

I apologize if I am asking too much or if these questions have already been answered. I am excited for a chance to contribute!

**Comment by pyu10055** [comment]
@AviKndr Thank you very much for volunteering to work on this. Here is my thinking on your questions:
1. You can view the shader program in GPU a function of filling the output tensor by applying functions on values from input tensor spaces. You will normally calculate the shape of the output tensor and provide the  function to do the calculation for each element. For example the [clip gpu implementation](https://github.com/tensorflow/tfjs-core/blob/master/src/kernels/webgl/clip_gpu.ts).


2. I will let @dsmilkov answer the back-propagation question, or you can implement the forward propagation first.

3. I believe the [test for crop_and _resize](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/kernels/crop_and_resize_op_test.cc)  has more output tensor tests.

4. usually just the current op examples in the [kernels/webgl](https://github.com/tensorflow/tfjs-core/blob/master/src/kernels/webgl/) directory.

**Comment by AviKndr** [comment]
@pyu10055 Thank you so much for the explanations. I have successfully implemented and tested all kernels on both backends. Will clean up my code and submit a pull request soon.

**Comment by dsmilkov** [comment]
Thanks @AviKndr for your contribution. Looking forward to your PR. +1 to @pyu10055's answers. Here are my responses as well:

1. Flat indexing for the CPU is the way to go. Thank you for offering to migrate the other resize ops too.

2. To keep the initial PR simpler and faster to review, how about dropping gradients for now, and adding them in a follow up PR. This way we can have that particular discussion on the PR that adds gradients itself.

3. Usually we manually run python tensorflow and copy paste the output to the js world. But if you find exact tests in c/c++ with values, that works too.

**Comment by AviKndr** [comment]
@dsmilkov. Thanks, I submitted the pull request! 

Btw, I found that in the original gradient implementation, they approximated the box gradients with 'nearest' using the kernel intended for 'bilinear' [image_grad.py lines 117-122](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/ops/image_grad.py) so that should not actually be an issue if the plan is to stick to the reference implementation.
